fix(generic): check local token expiry

becm opened the pull request:
Not checking for expired tokens triggers failures on first fetch/push after expiration.
Many Oauth2 implementations use JWT, where expiration time stamp is stored in structured data.

Fixes #268
Fixes #1408
Fixes #1784

becm commented:
I'd be open to split off the simple straightforward fix for saving the new `refresh_token`.

But the I'd very much like to have _any_ approach towards identifying local token expiration.
Alternative idea:
- the username is (for the generic provider, at the moment) unused.
- we can change the stored value to `OAUTH_USER:<expiration>` and strip the value before returning.

The last approach changing the API for the storage format (#1464) is now stale from a year.
Neither suggestions here would need additional code to be touched.
The `username` way is likely even _waaayyy_  more compact (string concat/split in ~2*3 locations).

becm commented:
Some major caveat found while trying to prototype a "<username>@<expiration>" approach:
API may **only** return `expires_in` for the new `access_token`, not an update `refresh_token`!

Relative time in protocol additionally indicates this is to be used as a transient element.
Committing that value to permanent storage (after calculating an absolute date) feels like a data flow violation.

For now definitely favor using structured token detection (transparent, backward-compatible, non-intrusive).

hickford commented on src/shared/Core/GenericHostProvider.cs:
```diff
@@ -125,6 +126,19 @@ public override async Task<ICredential> GenerateCredentialAsync(InputArguments i
 
+        public override async Task<ICredential> GetCredentialAsync(InputArguments input)
+        {
+            var credential = await base.GetCredentialAsync(input);
```
So depending on the host, the password is sometimes a JSON Web Token? I suggest to add this in a comment.

becm replied:
Addressed in comments for `auth_token` and `refresh_token` checks.

hickford commented on src/shared/Core/GenericHostProvider.cs:
```diff
@@ -125,6 +126,19 @@ public override async Task<ICredential> GenerateCredentialAsync(InputArguments i
+        public override async Task<ICredential> GetCredentialAsync(InputArguments input)
+        {
+            var credential = await base.GetCredentialAsync(input);
+            if (WebToken.IsExpiredToken(credential.Password))
```
For readability, I think the following would be clearer:

```suggestion
            if (WebToken.TryParse(credential.Password, out var token) && token.IsExpired())
```

becm replied:
Changed `TryParse` result to boolean and with new `out` parameter.
Obvious, but I just did not think of the _correct_ compact form for C#. ☹

lostmsu commented:
@mjcheetham @mpysson there've been no changes in this project for 5 months. Can I co-maintain it? Not sure if I can do it long term but I promise to go through the currently open pull requests at the very least. Better than nothing.

mjcheetham reviewed (commented):
What do you think about the merits of using the `JsonWebToken` type from the https://www.nuget.org/packages/Microsoft.IdentityModel.JsonWebTokens package?

becm commented:
@mjcheetham I'd prefer to apply the (optional) switch to a "proper Identity library" at a later date.
Doing this _by hand_ also has the benefit of only having code for the actually _needed_ operations.

The update to `.Net 10` will be a major change in dependencies anyway and we can _do it properly_ (if desired).
This will also be a good chance to replace the local Base64Url code (one centralized switch of dependencies).

In the current (revised) approach, external code should not be affected.

becm committed:
feat: add decode support to Base64Url converter

move definitions into private class scope

becm committed:
fix(generic): check expiry of access tokens

add token expiry check for credential `password` value
add generic StructuredToken class with expiration status property
include minimal JWT reader to decode and extract data

becm commented:
Removed the _expiration check_ from the `refresh_token`, as there is basically no benefit.
The existing code (and request) paths _should_ cope with and recover from using an expired token.

This should never cause issues in the _proper_ single well-behaved client case.

If an update to a `refresh_token` is missed (bad client behavior or parallel use of same client ID)
the expiry date of the local token becomes invalid without notice
(immediately or after some minutes, depending on server setup).

ryanbuck182 commented:
Let's get this merged!

becm committed:
fix(generic): check expiry of access tokens

add token expiry check for credential `password` value
add generic Token interface with expiration status
add minimal JsonWebToken reader to decode and extract data

becm commented:
New approach for `GCMv3`/`.net10+`:
- adjust to source tree restructure
- improve credential code flow and log messages
- extend interface for use of `authtype/credential` in Git credential
- avoid problematic approach for JSON deserializer
- split generic and JWT token classes
- be lenient on optional JWT elements
